specs/dwritecore/DWriteCore.md

```diff
@@ -1235,3 +2025,906 @@ DWRITE_BEGIN_INTERFACE(IDWriteFactory8, "202D44F0-84AC-4D6F-B8D0-704B27675CAA")
+defined in the included version of `dcommon.h`. This is accomplished as follows in `dwrite_3.h`:
+
+```c++
+/// <summary>
+/// DWRITE_GLYPH_IMAGE_FORMATS_COLR_PAINT_TREE image format represents a color glyph that can be
+/// rendered by drawing paint elements returned by IDWritePaintReader.
+/// </summary>
+#ifndef DWRITE_GLYPH_IMAGE_FORMATS_COLR_PAINT_TREE_DEFINED
+#define DWRITE_GLYPH_IMAGE_FORMATS_COLR_PAINT_TREE_DEFINED
+#define DWRITE_GLYPH_IMAGE_FORMATS_COLR_PAINT_TREE static_cast<DWRITE_GLYPH_IMAGE_FORMATS>(0x00000100)
+#endif
+```
+
+### DWRITE_PAINT_FEATURE_LEVEL Enumeration
+
+```c++
+/// <summary>
+/// Defines known feature level for use with the IDWritePaintReader interface and
+/// related APIs. A feature level represents a level of functionality. For example, it
+/// determines what DWRITE_PAINT_TYPE values might be returned.
+/// </summary>
+/// <remarks>
+/// See the DWRITE_PAINT_TYPE enumration for which paint types are required for each
+/// feature level.
+/// </remarks>
+enum DWRITE_PAINT_FEATURE_LEVEL
+{
+    /// <summary>
+    /// No paint API support.
+    /// </summary>
+    DWRITE_PAINT_FEATURE_LEVEL_NONE = 0,
+
+    /// <summary>
+    /// Specifies a level of functionality corresponding to OpenType COLR version 0.
+    /// </summary>
+    DWRITE_PAINT_FEATURE_LEVEL_COLR_V0 = 1,
+
+    /// <summary>
+    /// Specifies a level of functionality corresponding to OpenType COLR version 1.
+    /// </summary>
+    DWRITE_PAINT_FEATURE_LEVEL_COLR_V1 = 2
+};
+```
+
+### DWRITE_PAINT_ATTRIBUTES Enumeration
+
+```c++
+/// <summary>
+/// Combination of flags specifying attributes of a color glyph or of specific color values in
+/// a color glyph.
+/// </summary>
+enum DWRITE_PAINT_ATTRIBUTES
+{
+    DWRITE_PAINT_ATTRIBUTES_NONE = 0,
+
+    /// <summary>
+    /// Specifies that the color value (or any color value in the glyph) comes from the font's
+    /// color palette. This means the appearance may depend on the current palette index, which
+    /// may be important to clients that cache color glyphs.
+    /// </summary>
+    DWRITE_PAINT_ATTRIBUTES_USES_PALETTE = 0x01,
+
+    /// <summary>
+    /// Specifies that the color value (or any color value in the glyph) comes from the client-specified
+    /// text color. This means the appearance may depend on the text color, which may be important to
+    /// clients that cache color glyphs.
+    /// </summary>
+    DWRITE_PAINT_ATTRIBUTES_USES_TEXT_COLOR = 0x02
+};
+#ifdef DEFINE_ENUM_FLAG_OPERATORS
+DEFINE_ENUM_FLAG_OPERATORS(DWRITE_PAINT_ATTRIBUTES)
+#endif
+```
+
+### DWRITE_PAINT_COLOR Structure
+
+```c++
+/// <summary>
+/// Represents a color in a color glyph.
+/// </summary>
+struct DWRITE_PAINT_COLOR
+{
+    /// <summary>
+    /// Color value.
+    /// </summary>
+    DWRITE_COLOR_F value;
+
+    /// <summary>
+    /// Specifies the palette entry index if the color value comes from the font's color palette.
```

Comment:
Does the application have any insight into the font's actual color palette? In your examples, we always either use an index of 0, or we detect the forground brush palette index (0xffff) and use the foreground color. Are there ever any cases when an app would choose a different palette index?

Reply:
There are two kinds of indices. A font can have multiple palettes. A _color palette index_ determines which color palette is used and is specified by the client. A _palette entry index_ selects an entry within a palette and is specified by a record in the font (part of a paint record or gradient stop record or COLR v0 layer record).

Zero is a special color palette index, meaning the default palette. Other color palette indices depend on the font and have no standard meaning. The client specifies the color palette index as a parameter to `TranslateColorGlyphRun` or `IDWritePaintReader::SetColorPaletteIndex`.

0xFFFF is a special palette entry index, meaning the foreground brush. In all other cases, the color `value` is determined by using the paletteEntryIndex as an index into the client-specified color palette. The color's alpha value is then multiplied by the `alphaMultiplier`, which also comes from a color record in the font.

It is also possible for a client to define a custom palette instead of using a color palette from the font. CSS allows this. This is one reason `DWRITE_PAINT_COLOR` exposes the information it does, so a client with a custom palette or other custom behavior has the information it needs to compute the color value for a paint element on its own.

Reply:
Is it worth adding a paragraph or two basically saying what you describe here? As a reader I was a little unclear when or why you'd ever intentionally specify a palette/index other than 0 or 0xffff, and how you would understand what colors/palette that index would refer to. That is, if this is not already mentioned in some other part of the spec.

Reply:
I've updated the comments to DWRITE_PAINT_COLOR to explain some of the above.